### PR TITLE
Fix docs anchors across marker blocks

### PR DESCRIPTION
**Pull request by diegohaz**
## Motivation

The docs generator builds table-of-contents links for generated API reference blocks, but GitHub assigns heading anchors across the whole rendered README. When a README has handwritten headings or multiple generated marker blocks with the same heading slug, a per-block slugger can produce links that point to an earlier heading instead of the generated entry.

This is already a latent problem in `@ariakit/test`, where multiple API reference blocks can render entries with the same name. The script also computed member back-to-top links with a raw section-heading slug, so collisions with the section heading could desync those links from GitHub's anchor assignment too.

## Solution

Seed generated docs with the ATX Markdown headings that appear before the target marker, then use the same slugger for the section heading, table of contents, module headings, and member headings. The heading scan ignores fenced code blocks and HTML comments, normalizes simple inline Markdown links to their rendered text before slugging, and only considers headings before the marker so later document content cannot affect earlier generated anchors.

The docs freshness test now mirrors the injection path by computing the same preceding-heading seed for each marker block. Additional regression tests cover preceding versus later heading collisions, section-heading collisions, fenced code blocks, HTML comments, inline link headings, and setext-style blocks being ignored.

### CODE REVIEW

File: packages/ariakit-scripts/src/docs.ts [hunk 1]
```diff
@@ -670,15 +671,78 @@ function createSlugger() {
...
+      setextHeadingLines = [];
+      continue;
+    }
+
+    const setextHeadingLine = line.match(/^ {0,3}(\S.*)$/)?.[1]?.trim();
+    if (setextHeadingLine) {
+      setextHeadingLines.push(setextHeadingLine);
+    } else {
+      setextHeadingLines = [];
+    }
```
Comment (diegohaz) [inline packages/ariakit-scripts/src/docs.ts]: **Consider dropping Setext support rather than hardening it — it protects no current readme and the freshness test can't catch a wrong seed.**

The Setext branch accumulates *any* run of non-blank lines and emits them as a heading when a `=`/`-` underline follows. As CodeRabbit and Codex already noted, that mis-seeds list items, blockquotes, and thematic breaks (`- item` / `---`) as headings. Two extra facts make this worth a second look:

1. **No committed readme uses Setext headings.** I grepped all 7 readmes with docs markers (`ariakit-test`, `ariakit-store`, `ariakit-solid-store`, `ariakit-react-store`, `ariakit-utils`, `ariakit-react-utils`, `ariakit-solid-utils`) — every heading before every marker is ATX (`#`). So this whole branch (and its two new tests) guards a case that doesn't exist in the repo, while introducing the false-positive class above.

2. **The freshness test cannot detect a bad seed.** `"$dir readme docs are up to date"` regenerates with `getPrecedingHeadings(readme, start)` and compares to the committed block. A committed readme was *also* generated with the same `getPrecedingHeadings`, so a mis-seeded anchor agrees with itself and the test passes — even though GitHub's real anchor differs and the generated TOC/back-to-top link is broken. These edge cases are therefore silent, not just rare.

Given that, the lowest-complexity option is to seed ATX headings only and drop the Setext accumulator entirely. A future Setext heading before a marker would then simply not be seeded (an under-suffix, which only breaks if a generated entry happens to collide) — a strictly smaller and rarer failure mode than the current over-seeding from list items, and far less code than hand-rolling a correct Setext + thematic-break + list disambiguation.

If you'd rather keep Setext (e.g. future-proofing), then the bot threads' tightening is the minimum, but note you'd also need HTML-comment skipping and link-text normalization (`## [foo](#x)` → `foo`) to actually mirror GitHub — i.e. you're incrementally reimplementing a CommonMark parser by hand for inputs that don't occur today.

**Review by diegohaz — COMMENTED** [review]
### Review summary

**Scope:** Initial review of the full diff (merge base `f4cc781` → head `c634954`), expanded into `packages/ariakit-scripts/src/docs.ts`, the new tests in `docs.test.ts`, and all 7 readmes that use docs markers.

**Verdict:** The fix is correct for every readme in the repo and the approach is sound. The shared slugger now seeds preceding headings, the section heading, the TOC, module headings, and member back-to-top links from one ordered occurrence map, which matches GitHub's top-down anchor assignment. One inline comment, a simplification opportunity (not a blocker).

**What I verified locally:**
- New anchor tests pass: `uses/ignores preceding readme headings`, `disambiguates entries and back-to-top links from section headings`, `ignores fenced code`, `collects setext headings` (run individually, all green).
- **No regression risk for current readmes.** The `topAnchor = slugifyHeading(heading)` → `slug(heading)` change only differs when a preceding heading collides with the section-heading slug. None of the 7 readmes have a preceding heading that slugs to `api-reference` / `react-api-reference` / `playwright-api-reference` (they precede only `Contents` / `Installation` / `Usage` / the package H1), and no `readme.md` changed in this PR, so output is byte-stable.
- Traced the slugger call order (preceding → section heading → TOC group titles/entries) against the rendered body order (`##` heading → `### module` → `#### member`); they match, including the module path exercised by `@ariakit/utils`. Member headings render the name in backticks and `slugifyHeading` strips them, so seeds line up with GitHub.
- The `markerIndex === -1 → whole readme` fallback is correct for first-time injection, where the block is appended after all existing headings.
- The fence tracker correctly handles nested/longer fences, info strings on closing lines, and resets a pending Setext candidate when a fence opens.

**Findings:**
- *Simplification (inline, non-blocking):* the Setext branch is the source of the false-positive seeding that CodeRabbit and Codex flagged, **and** I confirmed no committed readme uses Setext headings, so it guards a case that doesn't occur while adding that risk. I also note the freshness test can't catch a bad seed (round-trip regeneration agrees with the committed file even when GitHub's anchor differs), so these edge cases are silent rather than merely rare. Recommended dropping Setext seeding (ATX-only) over hand-rolling a correct Setext + list + thematic-break + HTML-comment + link-text parser.
- The existing bot threads (Setext from list items, HTML-comment headings, link-text normalization) are all legitimate but unexercised by today's readmes; folded into the inline comment rather than re-posted.

**Not fully validated:** The `"$dir readme docs are up to date"` freshness test fails locally with multi-line signature re-wrapping (e.g. `waitFor`), but it fails identically on `origin/main`, so it's a pre-existing local `oxfmt`/React-types environment mismatch, not introduced by this PR. I could not exercise the committed-readme regeneration path end-to-end locally because of it.

File: packages/ariakit-scripts/src/docs.ts [hunk 2]
```diff
@@ -670,15 +671,78 @@ function createSlugger() {
...
+    const heading = headingMatch?.[1];
+    if (heading) {
+      headings.push(heading);
+      setextHeadingLines = [];
+      continue;
+    }
+
+    const setextHeadingLine = line.match(/^ {0,3}(\S.*)$/)?.[1]?.trim();
+    if (setextHeadingLine) {
+      setextHeadingLines.push(setextHeadingLine);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts]: Confirmed this is real: the candidate accumulator at L727 pushes any `\S.*` line into `setextHeadingLines`, so a list run, blockquote, or thematic break followed by `=`/`-` (`- item` then `---`) seeds a phantom heading that GitHub never renders.

Two notes on priority before applying the `isSetextCandidate` tightening:
- I grepped all 7 readmes with docs markers (`ariakit-test`, `ariakit-store`, `ariakit-solid-store`, `ariakit-react-store`, `ariakit-utils`, `ariakit-react-utils`, `ariakit-solid-utils`) — none use Setext headings anywhere, so this branch currently guards no real input while introducing exactly this false positive.
- The freshness test can't catch a bad seed: it regenerates with the same `getPrecedingHeadings` and compares to the committed block, so a mis-seeded anchor agrees with itself even though GitHub's real anchor differs. These cases are silent, not just rare.

Your `isSetextCandidate` regex is the right minimum if Setext stays. I'd also weigh dropping the Setext branch entirely (ATX covers every current readme) — full reasoning in https://github.com/ariakit/ariakit/pull/6499#discussion_r3477227552.

File: packages/ariakit-scripts/src/docs.ts [hunk 3]
```diff
@@ -670,15 +671,78 @@ function createSlugger() {
...
+    if (setextMatch && setextHeadingLines.length) {
+      headings.push(setextHeadingLines.join(" "));
+      setextHeadingLines = [];
+      continue;
+    }
+
+    const headingMatch = line.match(/^ {0,3}#{1,6}\s+(.+?)\s*#*\s*$/);
+    const heading = headingMatch?.[1];
+    if (heading) {
+      headings.push(heading);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts:737]: Verified: `slugifyHeading("[foo](#details)")` strips the brackets/parens/`#` and yields `foodetails`, whereas GitHub anchors the *rendered* link text (`foo`). So a preceding `## [foo](#details)` would seed `foodetails` instead of `foo`, and a generated `foo` entry would be linked `#foo` while GitHub assigns it `#foo-1`. Real.

Worth noting this isn't Setext-specific — it hits ATX headings too (`## [foo](#x)` above a marker). It's currently unexercised: every heading before a marker in the 7 readmes is either plain text (section headings) or an inline-code member name (`### ` + backticked name), and `slugifyHeading` already strips backticks so those line up with GitHub. No link-text heading exists before a marker today.

Consolidated with the other scanner edge cases in https://github.com/ariakit/ariakit/pull/6499#discussion_r3477227552.

[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts:737]: Confirmed: a `## X` line inside a multi-line HTML comment matches `headingMatch` and gets seeded even though GitHub renders nothing for it (and would then over-suffix a later real `X` heading).

Scope check so this isn't conflated with the marker handling: the single-line docs markers (`<!-- ariakit-docs:start -->`) contain no heading syntax and are fine, and the genuine `##`/`###` headings inside *earlier generated blocks* SHOULD be seeded — they're real rendered headings (this is how the playwright block's `query` correctly resolves to `#query-1`). The gap is strictly heading-like lines inside HTML comments, which no current readme has before a marker.

Real but unexercised; folded into the broader scanner-vs-CommonMark point in https://github.com/ariakit/ariakit/pull/6499#discussion_r3477227552.

File: packages/ariakit-scripts/src/docs.ts [hunk 4]
```diff
@@ -670,15 +671,78 @@ function createSlugger() {
...
+        setextHeadingLines = [];
+      }
+      continue;
+    }
+
+    if (fence) continue;
+
+    const setextMatch = line.match(/^ {0,3}(=+|-+)\s*$/);
+    if (setextMatch && setextHeadingLines.length) {
+      headings.push(setextHeadingLines.join(" "));
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts]: Same root cause as CodeRabbit's thread on L729 — the candidate accumulator treats every non-blank line as Setext paragraph text, so `- item one` / `- item two` / `---` emits a phantom `- item one - item two` heading. Confirmed.

Unexercised today: none of the 7 marker readmes use Setext headings at all, and the freshness test can't surface it (it regenerates with the same `getPrecedingHeadings` it's checking). Consolidated take, plus a drop-Setext-entirely alternative, in https://github.com/ariakit/ariakit/pull/6499#discussion_r3477227552.

File: packages/ariakit-scripts/src/docs.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts]: Fixed in 08818df68 - removed Setext heading scanning from `getPrecedingHeadings` and added coverage for setext-style/list blocks being ignored, so list items plus thematic breaks no longer seed phantom headings.

File: packages/ariakit-scripts/src/docs.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts:737]: Fixed in 08818df68 - `slugifyHeading` now normalizes simple inline Markdown link/image syntax to rendered text before seeding slugs, with an integration test for `## [foo](#details)`.

[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts:737]: Fixed in 08818df68 - `getPrecedingHeadings` now strips HTML comments outside fenced code before matching ATX headings, with coverage for multi-line and inline comments.

File: packages/ariakit-scripts/src/docs.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts]: Fixed in 08818df68 - dropped Setext parsing and added a regression test for `- item` followed by `---` so it is not seeded.

File: packages/ariakit-scripts/src/docs.ts (hunk 1 above)
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts]: Fixed in 08818df68 - dropped Setext support, made the helper ATX-only, removed the Setext-positive test, and updated the PR description to match.

**Comment by diegohaz** [comment]
**Fixed**

- 08818df68 ([reply](https://github.com/ariakit/ariakit/pull/6499#discussion_r3477371224)): Removed Setext heading scanning and added coverage for setext-style/list blocks being ignored, so list items plus thematic breaks no longer seed phantom headings.
- 08818df68 ([reply](https://github.com/ariakit/ariakit/pull/6499#discussion_r3477371778)): Normalized simple inline Markdown link/image syntax to rendered text before slugging, with an integration test for `## [foo](#details)`.
- 08818df68 ([reply](https://github.com/ariakit/ariakit/pull/6499#discussion_r3477372335)): Stripped HTML comments outside fenced code before matching ATX headings, with coverage for multi-line and inline comments.
- 08818df68 ([reply](https://github.com/ariakit/ariakit/pull/6499#discussion_r3477373360)): Dropped Setext parsing and added a regression test for `- item` followed by `---` so it is not seeded.
- 08818df68 ([reply](https://github.com/ariakit/ariakit/pull/6499#discussion_r3477374120)): Made the helper ATX-only, removed the Setext-positive test, and updated the PR description to match.

File: packages/ariakit-scripts/src/docs.test.ts [hunk 5]
```diff
@@ -238,6 +239,215 @@ test("probes past taken suffixes when building contents links", () => {
...
+    "==============",
+    "",
+    "- item",
+    "---",
+    "",
+    start,
+    "",
+  ].join("\n");
+
+  expect(getPrecedingHeadings(readme, start)).toEqual(["Test"]);
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.test.ts:448]: Re-reviewed on `08818df`. The factual claim checks out — GitHub post-processes anchors onto Setext headings too, so dropping Setext seeding is a real divergence when a README has a genuine `Title` / `===` heading before a marker.

That said, I'd ship this as-is. The under-seed was the explicit fork in the original recommendation: *drop Setext, or implement it correctly* — the naive accumulator was strictly worse because it over-seeded list items and thematic breaks. The author took the drop path and documented it (the `getPrecedingHeadings` JSDoc now says "ATX Markdown headings" and the PR description matches). Specifically:
- No committed readme uses Setext, so output is byte-identical for every real input today (verified across all 7 marker readmes — their preceding content has no Setext, linked headings, or non-marker comments, so the removed branch was a no-op there).
- The residual risk is narrow: a *genuine* Setext heading before a marker **and** a generated entry/module that slugs to the same value — only then does the missing seed yield `#foo` instead of `#foo-1`.
- A correct re-introduction isn't a one-liner. `Title`/`---` is a Setext H2, but `- item`/`---` is list + thematic break, `Title`⏎⏎`---` is paragraph + break, Setext paragraphs can span multiple lines, and a lone `-` underline is ambiguous. The old accumulator got exactly these wrong. If Setext support is wanted later, it needs a guarded single-paragraph detector (reject lines starting with `-+*`, `>`, `N.`), not the old accumulator.

So: a real but rare gap, now an explicit and documented scope limitation. I would not block on it.

File: packages/ariakit-scripts/src/docs.ts [hunk 6]
```diff
@@ -664,15 +666,89 @@ function createSlugger() {
...
+        fence = { character, length };
+      }
+      continue;
+    }
+
+    if (fence) continue;
+
+    const headingMatch = lineWithoutComments.match(
+      /^ {0,3}#{1,6}\s+(.+?)\s*#*\s*$/,
+    );
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts:734]: Same point as CodeRabbit's thread on `docs.test.ts:448`, and I verified it: GitHub does anchor Setext headings, so this is a real (if rare) under-seed. My take and the trade-off analysis are on that thread — net, I'd ship the ATX-only scope as a documented limitation rather than re-add a Setext detector, since the naive accumulator over-seeded lists/thematic breaks and a correct one is non-trivial. Not a blocker.

File: packages/ariakit-scripts/src/docs.ts [hunk 7]
```diff
@@ -635,6 +636,7 @@ function renderEntry(entry: ApiEntry, heading: string, topAnchor: string) {
 
 function slugifyHeading(heading: string) {
   return heading
+    .replace(/!?\[([^\]]*)\]\([^)]+\)/g, "$1")
```
[the remark this answers is not in the document]
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts:639]: Confirmed on `08818df`: `slugifyHeading("[foo][details]")` → `foodetails` (the new regex only matches inline `](url)`, not full-reference `][label]`), whereas GitHub renders the reference text `foo` → `#foo`. Real.

But this is the extreme tail: it needs a *full*-reference linked heading (`[foo][label]` — the shortcut form `[foo]` already slugs to `foo`, so it's fine) with a matching link definition, before a marker, plus a colliding generated name. The commit scopes itself to "simple inline link/image syntax," and no readme has any linked heading before a marker. I'd leave this as a documented limitation rather than grow the regex toward a reference-definition resolver. Not a blocker.

File: packages/ariakit-scripts/src/docs.ts (hunk 1 above)
Comment (diegohaz) [reply packages/ariakit-scripts/src/docs.ts]: Verified the fix on `08818df`: Setext scanning removed (the helper is now ATX-only per its JSDoc), HTML comments stripped outside fenced code, and `slugifyHeading` normalizes inline link/image text before slugging. I ran the edge-case tests (link-text, HTML-comment, ignored-setext, fenced-code, preceding/later collisions, section-heading disambiguation) — all green — and adversarially probed `getPrecedingHeadings` (multi-line/inline comments, `<!---->`, `## C <!-- x --> D` → `c-d`, multi-link headings → `foo-and-bar`); all behave. Output is byte-stable for the 7 marker readmes since their preceding content exercises none of the changed paths.

The only open thread is the deliberate Setext under-seed (now on the `docs.test.ts:448` thread) — I think it's fine to ship as a documented limitation.

**Review by diegohaz — COMMENTED** [review]
### Re-review summary

**Re-review**, incremental baseline = my last reviewed SHA `c634954`. Since then the branch was rebased onto latest `main` (picking up #6485's `formatDeclarationText`, unrelated) and one new commit was added: `08818df Handle docs heading scanner edge cases`. This review covers that commit plus the new bot threads it triggered.

**What the new commit does (and it directly addresses the prior findings):**
- Drops the Setext branch entirely; `getPrecedingHeadings` is now ATX-only (JSDoc + PR description updated to match).
- Strips HTML comments (inline and multi-line) outside fenced code before matching headings.
- Normalizes simple inline link/image syntax in `slugifyHeading` (`## [foo](#x)` → `foo`).
- Test changes: inverts the Setext test to assert Setext-style/list blocks are ignored, adds link-text and HTML-comment coverage.

**Verification (on `08818df`):**
- The edge-case + anchor tests pass (link-text, HTML-comment, ignored-setext, fenced-code, preceding/later collisions, section-heading disambiguation).
- Adversarially probed `getPrecedingHeadings`: multi-line/inline comments, empty `<!---->`, `## C <!-- x --> D` → `c-d`, multi-link headings → `foo-and-bar` — all correct, and `stripHtmlComments` provably terminates (each loop iteration returns or strictly shortens the line).
- **No regression:** the only HTML comments in the 7 marker readmes are the single-line docs markers, no heading shares a line with a comment, and none use Setext or linked headings — so all three changed code paths are no-ops for every committed readme, and output stays byte-stable (consistent with no `readme.md` in the diff).

**Adjudication of the new bot findings (all non-blocking):**
- *Setext under-seed* (CodeRabbit `docs.test.ts:448` + Codex): factually correct — GitHub does anchor Setext headings, so dropping them is a real divergence. But it was the explicit trade-off in the prior recommendation (the naive accumulator over-seeded lists/thematic breaks; a correct Setext detector is non-trivial given the `---` list/break/heading ambiguity and multi-line paragraphs). With zero Setext usage in any readme and the limitation now documented, I'd ship as-is. Replied on both threads.
- *Reference-style links* (Codex): `[foo][label]` isn't normalized (only inline `](url)` is). Real but the extreme tail — needs a full-reference linked heading + definition + a colliding generated name; the commit scopes to "simple inline" and no readme has linked headings before a marker. Acceptable as a documented limitation. Replied on the thread.

**Verdict:** The new commit cleanly resolves the prior findings; no new actionable defects. The remaining open item (Setext) is a deliberate, documented scope limitation that is correct for every readme in the repo. LGTM from my side.

**Not fully validated:** the `"…docs are up to date"` freshness test still fails locally on signature re-wrapping — reproduces identically on `origin/main`, so it's a pre-existing local `oxfmt`/React-types env mismatch, not from this PR.
